### PR TITLE
build: make libuv buildable with vs2017

### PR DESCRIPTION
**Pull request by refack**
Re: #1283 

This is the minimal change needed to get this to build on VS2017, it assumes the dev has run `vsvarsall.bat` (i.e. set-up the dev environment, and chose target and host arch) So to get it to build on appveyor you need to add to the YAML:
```
before_build:
  - cmd: "C:\Program Files (x86)\Microsoft Visual Studio 15.0\VC\Auxiliary\Build\vcvarsall.bat amd64"
```

I could add another PR with VS2017 detection code...

### CODE REVIEW

**Comment by saghul** [comment]
How does this relate to #1283 ? Can you provide some more context?

**Comment by refack** [comment]
- This change will make `libuv` buildable on windows with VS2017.
- It does not install or setup VS2017 (as #1283 [tries to do](https://github.com/ctaggart/libuv/blob/4d4ca864d0a0312c90f1a07d77a66f770801873e/scripts/Add-VisualStudioPackages.ps1)).
- I did not change `appveyor.yml`. I put the needed change in the description. I could add it if you want.
- This does not include VS2017 detection logic (I usually use [`msvs-com-helper`](https://github.com/node4good/msvs-com-helper))
- #1283 has a [script that manipulates](https://github.com/ctaggart/libuv/blob/4d4ca864d0a0312c90f1a07d77a66f770801873e/scripts/Update-ProjectFilesToVS2017.ps1) the `GYP` output. This PR makes that redundant.

File: appveyor.yml [hunk 1]
```diff
@@ -1,10 +1,13 @@
 version: v1.11.0.build{build}
 
+image:
```
Comment (saghul) [inline appveyor.yml]: The changes here look like will make appveyor only use VS 2017, is that so?

File: appveyor.yml [hunk 2]
```diff
@@ -1,10 +1,13 @@
...
+  - Visual Studio 2017
+
 install:
   - cinst -y nsis
 
 matrix:
-  fast_finish: true
+  fast_finish: false
```
Comment (saghul) [inline appveyor.yml]: what is this supposed to do?

**Comment by refack** [comment]
Green on Appveyor [Build status](https://ci.appveyor.com/project/refack/libuv/branch/vs2017)

File: gyp_uv.py [hunk 3]
```diff
@@ -34,37 +44,47 @@ def host_arch():
   return machine  # Return as-is and hope for the best.
 
 
-def run_gyp(args):
-  rc = gyp.main(args)
-  if rc != 0:
-    print('Error running GYP')
-    sys.exit(rc)
+def config_vs2017(config_fn):
```
Comment (saghul) [inline gyp_uv.py]: can we ship this file instead of writing it every time?

File: gyp_uv.py [hunk 4]
```diff
@@ -34,37 +44,47 @@ def host_arch():
...
 
   # GYP bug.
   # On msvs it will crash if it gets an absolute path.
   # On Mac/make it will crash if it doesn't get an absolute path.
   if sys.platform == 'win32':
-    args.append(os.path.join(uv_root, 'uv.gyp'))
-    common_fn  = os.path.join(uv_root, 'common.gypi')
-    options_fn = os.path.join(uv_root, 'options.gypi')
-    # we force vs 2010 over 2008 which would otherwise be the default for gyp
-    if not os.environ.get('GYP_MSVS_VERSION'):
```
Comment (saghul) [inline gyp_uv.py]: why is this removed?

File: gyp_uv.py [hunk 5]
```diff
@@ -34,37 +44,47 @@ def host_arch():
...
+    arg_path = uv_root
   else:
-    args.append(os.path.join(os.path.abspath(uv_root), 'uv.gyp'))
-    common_fn  = os.path.join(os.path.abspath(uv_root), 'common.gypi')
-    options_fn = os.path.join(os.path.abspath(uv_root), 'options.gypi')
+    arg_path = os.path.abspath(uv_root)
+
+
+  args.append(os.path.join(arg_path, 'uv.gyp'))
+  common_fn  = os.path.join(arg_path, 'common.gypi')
```
Comment (saghul) [inline gyp_uv.py]: notice the difference in absolute vs relative paths, I don't recall if that was intentional or not, but it looks like it.

File: gyp_uv.py [hunk 6]
```diff
@@ -89,6 +109,13 @@ def run_gyp(args):
...
-  print(gyp_args)
-  run_gyp(gyp_args)
+  print(args)
+  rc = gyp.main(args)
+  if rc != 0:
+    print('Error running GYP')
+    sys.exit(rc)
+
+
+if __name__ == '__main__':
```
Comment (saghul) [inline gyp_uv.py]: this is unnecessary, this file is never used as a module

File: appveyor.yml (hunk 1 above)
Comment (refack) [inline appveyor.yml]: Yes. I can add it to the matrix.

File: appveyor.yml (hunk 2 above)
Comment (refack) [inline appveyor.yml]: `true` will stop the build on first failure (making the `allow_failures:` lines contradictory)

File: gyp_uv.py (hunk 3 above)
Comment (refack) [inline gyp_uv.py]: SDK version depends on dev ENV. (toolset might change in the future as well)

File: gyp_uv.py (hunk 4 above)
Comment (refack) [inline gyp_uv.py]: I parameterized the `base_path` and merged the two cases.

File: gyp_uv.py (hunk 5 above)
Comment (refack) [inline gyp_uv.py]: It is intentional as stated in [line 64](https://github.com/refack/libuv/blob/c7b745035369966c639456dae52b642306b1340c/gyp_uv.py#L64) but could be parameterized.

File: gyp_uv.py (hunk 6 above)
Comment (refack) [inline gyp_uv.py]: Yes, but still cleaner 😉
I'm pushing for the same change in `nodejs`

File: appveyor.yml (hunk 1 above)
Comment (saghul) [inline appveyor.yml]: Well, your PR should *add*, not replace.

File: gyp_uv.py (hunk 6 above)
Comment (saghul) [inline gyp_uv.py]: Please remove it.

File: gyp_uv.py (hunk 4 above)
Comment (saghul) [inline gyp_uv.py]: I mean setting the default GYP_MSVS_VERSION

File: gyp_uv.py [hunk 7]
```diff
@@ -34,37 +44,47 @@ def host_arch():
...
-  rc = gyp.main(args)
-  if rc != 0:
-    print('Error running GYP')
-    sys.exit(rc)
+def config_vs2017(config_fn):
+  if os.environ.get('VisualStudioVersion') != '15.0': return
+  o = dict()
+  td = o['target_defaults'] = dict()
+  try:
+    td['msbuild_toolset'] = 'v141'
```
Comment (saghul) [inline gyp_uv.py]: why do we want to do this only if the version if != than 15.0 ?

**Comment by saghul** [comment]
I'm at a loss here. Can you please walk me through what you are trying to accomplish here? Yes, MSVS 2017 support, but I need to know the details: why is a new file required? Why are we removing checks we used before (which I know people rely upon), etc.

File: appveyor.yml (hunk 1 above)
Comment (refack) [inline appveyor.yml]: I agree on principle... was looking for fast feedback from AppVeyor (and from you).

File: gyp_uv.py (hunk 4 above)
Comment (refack) [inline gyp_uv.py]: Ohh, yea, that does requires an explanation... I was thinking it's not the place to set a default.
On Windows `GYP_MSVS_VERSION` comes from vcbuild.bat, but indeed I didn't check the posix flow, my bad.

File: gyp_uv.py (hunk 6 above)
Comment (refack) [inline gyp_uv.py]: Ok.

File: gyp_uv.py (hunk 7 above)
Comment (refack) [inline gyp_uv.py]: `os.environ.get('VisualStudioVersion') != '15.0':`  do nothing new (i.e. `return`)
I didn't see any `target_defaults` stuff, so I added it only in the VSver 15.0 case.
See my general comment.

**Comment by refack** [comment]
@saghul I'll revamp this as per your comments.
Currently it has some hardcoded values such as `os.environ.get('VisualStudioVersion') == '15.0'` => `td['msbuild_toolset'] = 'v141'`
If you want I can add the COM interrogator that is a little bit more future proof and can deduce this value.

**Comment by refack** [comment]
P.S. I have a heavy hand and I like to slash code I don't like, then let the CI keep me in check ;)

File: gyp_uv.py (hunk 4 above)
Comment (refack) [inline gyp_uv.py]: Bottom line `GYP_MSVS_VERSION` comes from detection code in `vcbuild.bat`
IMHO here it's dropped out of the blue.

**Comment by saghul** [comment]
> Currently it has some hardcoded values such as os.environ.get('VisualStudioVersion') == '15.0' => td['msbuild_toolset'] = 'v141'
> If you want I can add the COM interrogator that is a little bit more future proof and can deduce this value.

How is this supposed to work with MSVS 2010 or SDK 9.1, both of which we still support?

File: appveyor.yml (hunk 1 above)
Comment (refack) [inline appveyor.yml]: Fixed

**Comment by refack** [comment]
Fixed by your request:
* AppVeyor will do a 2x2 matrix (2015,2017)x(86, 84)
* `gyp_uv.py` will add a default to `GYP_MSVS_VERSION = 2010`. I just added some explanation to the comment
* un-refactored gyp_uv.gyp to be more module like

> How is this supposed to work with MSVS 2010 or SDK 9.1, both of which we still support?

using the same flow as before... the function will become a no-op, => config.gypi will not be generated, so it won't be added as an arg to `GYP`

**Comment by refack** [comment]
[Green again on the 2x2](https://ci.appveyor.com/project/refack/libuv/build/v1.11.0.build17)
@saghul BTW: If you want I can set AppVeyor to check on:
* Visual Studio 2008
* Visual Studio 2010
* Visual Studio 2012
* Visual Studio 2013
* Visual Studio 2015
* Visual Studio 2017
And I think you need to choose the one version that will make a release file (and not waste time packing them on all)

**Comment by refack** [comment]
@saghul you like the new changes?

**Comment by refack** [comment]
@saghul ping?

File: gyp_uv.py [hunk 8]
```diff
@@ -34,6 +44,19 @@ def host_arch():
...
+  # for now we only have config only for VS2017 a.k.a 15.0
+  # we dynamically generate the file because os.environ['WindowsSDKVersion']
+  # is system dependant
+  if os.environ.get('VisualStudioVersion') == '15.0':
+    o = dict()
+    td = o['target_defaults'] = dict()
+    td['msbuild_toolset'] = 'v141'
+    sdk_ver = os.environ['WindowsSDKVersion'].rstrip("\\")
+    td['msvs_windows_target_platform_version'] = sdk_ver
+    write(config_fn, o)
```
Comment (saghul) [inline gyp_uv.py]: please inline the write function here

File: gyp_uv.py [hunk 9]
```diff
@@ -34,6 +44,19 @@ def host_arch():
...
 
+def generate_config_gypi(config_fn):
+  # for now we only have config only for VS2017 a.k.a 15.0
+  # we dynamically generate the file because os.environ['WindowsSDKVersion']
+  # is system dependant
+  if os.environ.get('VisualStudioVersion') == '15.0':
+    o = dict()
+    td = o['target_defaults'] = dict()
+    td['msbuild_toolset'] = 'v141'
+    sdk_ver = os.environ['WindowsSDKVersion'].rstrip("\\")
```
Comment (saghul) [inline gyp_uv.py]: use single quotes

File: gyp_uv.py [hunk 10]
```diff
@@ -34,6 +44,19 @@ def host_arch():
   return machine  # Return as-is and hope for the best.
 
 
+def generate_config_gypi(config_fn):
+  # for now we only have config only for VS2017 a.k.a 15.0
+  # we dynamically generate the file because os.environ['WindowsSDKVersion']
+  # is system dependant
+  if os.environ.get('VisualStudioVersion') == '15.0':
+    o = dict()
```
Comment (saghul) [inline gyp_uv.py]: please, use a more descriptive variable name

File: gyp_uv.py [hunk 11]
```diff
@@ -24,6 +25,15 @@
   sys.exit(42)
 
 
+def write(filename, data):
+  filename = os.path.join(uv_root, filename)
+  print('creating %s' % filename)
+  content = pprint.pformat(data, indent=2) + '\n'
+  print (content)
```
Comment (saghul) [inline gyp_uv.py]: I'd rather not print the file

Comment (refack) [inline gyp_uv.py]: Just copied from `nodejs`. Maybe `if verbose`?

File: gyp_uv.py (hunk 8 above)
Comment (refack) [inline gyp_uv.py]: 😲

**Comment by refack** [comment]
@saghul changes addressed.

**Comment by refack** [comment]
Also I could tweak appveyor.yml in the manner I described in https://github.com/libuv/libuv/pull/1284#issuecomment-291130728

**Comment by refack** [comment]
Good news, VS2017 [landed in `GYP`]( https://chromium-review.googlesource.com/c/456360/)
PR updated, removed code to "trick" `GYP` to work.
@saghul hope you like this better.

File: vcbuild.bat [hunk 12]
```diff
@@ -41,6 +41,16 @@ shift
 goto next-arg
 :args-done
 
+@rem Look for a set-up Visual Studio 2017 environment
+:vs-set-2017
```
Comment (saghul) [inline vcbuild.bat]: this label is unused

File: vcbuild.bat [hunk 13]
```diff
@@ -41,6 +41,16 @@ shift
...
+if not defined VisualStudioVersion if not "%VisualStudioVersion%"=="15.0" goto :no-vs2017-env
+:found_vs2017
+echo Found MSVS 2017
```
Comment (saghul) [inline vcbuild.bat]: Please use the same wording as with other versions

File: vcbuild.bat [hunk 14]
```diff
@@ -41,6 +41,16 @@ shift
 goto next-arg
 :args-done
 
+@rem Look for a set-up Visual Studio 2017 environment
+:vs-set-2017
+if not defined VisualStudioVersion if not "%VisualStudioVersion%"=="15.0" goto :no-vs2017-env
+:found_vs2017
```
Comment (saghul) [inline vcbuild.bat]: unused label

File: gyp_uv.py [hunk 15]
```diff
@@ -48,16 +48,17 @@ def run_gyp(args):
   # On msvs it will crash if it gets an absolute path.
   # On Mac/make it will crash if it doesn't get an absolute path.
   if sys.platform == 'win32':
-    args.append(os.path.join(uv_root, 'uv.gyp'))
-    common_fn  = os.path.join(uv_root, 'common.gypi')
-    options_fn = os.path.join(uv_root, 'options.gypi')
-    # we force vs 2010 over 2008 which would otherwise be the default for gyp
+    arg_path = uv_root
+    # we force vs 2010 over 2008 which would otherwise be the default for gyp,
+	# since vs2008 uses old format `.vcproj` files
```
Comment (saghul) [inline gyp_uv.py]: indent looks off here

File: gyp_uv.py [hunk 16]
```diff
@@ -48,16 +48,17 @@ def run_gyp(args):
...
       os.environ['GYP_MSVS_VERSION'] = '2010'
   else:
-    args.append(os.path.join(os.path.abspath(uv_root), 'uv.gyp'))
-    common_fn  = os.path.join(os.path.abspath(uv_root), 'common.gypi')
-    options_fn = os.path.join(os.path.abspath(uv_root), 'options.gypi')
+    arg_path = os.path.abspath(uv_root)
+
+  args.append(os.path.join(arg_path, 'uv.gyp'))
+  common_fn  = os.path.join(arg_path, 'common.gypi')
+  options_fn = os.path.join(arg_path, 'options.gypi')
```
Comment (saghul) [inline gyp_uv.py]: please put these changes in a separate commit, since they are  no longer required for MSVS 2017 support.

File: appveyor.yml [hunk 17]
```diff
@@ -1,28 +1,51 @@
 version: v1.11.0.build{build}
 
+image:
```
Comment (saghul) [inline appveyor.yml]: I'm not sure we want this. AppVeyor is just some experimental builds we distribute. Can you please drop these changes and let's discuss them in a follow-up PR, since they are not related to getting the build going.

**Comment by refack** [comment]
@saghul Done. PTAL.

**Comment by saghul** [comment]
Nice and simple! I like it! Now, is this enough to build it using MSVC 2017 or is there something extra needed?

**Comment by refack** [comment]
Make sure the script pulls the new `GYP` (delete old downloaded ver). and that the dev run `vcvarsall.bat` before.
P.S. host-arch and target-arch are deduced from ENV

**Comment by saghul** [comment]
> and that the dev run vcvarsall.bat before.

Why can't we have this built in, as we do for other MSVC versions? (sorry if I asked before, this has changed so much over time I can't keep track)

**Comment by refack** [comment]
> Why can't we have this built in, as we do for other MSVC versions? (sorry if I asked before, this has changed so much over time I can't keep track)

MS changed the way to find the installed instance of VS2017, they remove `VS***COMNTOOLS` env var, and now you need to call a COM server.
I have a tool that does that with only builtin windows tools https://github.com/node4good/msvs-com-helper, I'm working on [landing it in `node` ](https://github.com/nodejs/node/pull/11852/files#diff-f1d059c92a388e475daa1fbc2a88aa3aR150), but it's a little bit cumbersome.
If you want it I can integrate it (it landed in `node-gyp`, `Boost/build`, and soon in `node` ;)

**Comment by refack** [comment]
Also @indutny took it into `gyp.js`

**Comment by refack** [comment]
@saghul a bit of good new, MS made it simpler to detect the VS version, but it will take some time until everyone updates, for we can detect only in some of the cases (falling back to the user calling `vcvarsall.bat` by themselves)

**Comment by refack** [comment]
@bnoordhuis I'm working on adding the detection mechanism @saghul asked for, now that MS made it simpler again to detect VS2017.

**Comment by refack** [comment]
Re the CI: it's really a regression test, since we don't have VS2017 workers...
P.S. Any idea why is `not ok 240 - fs_chown` failing?

**Comment by saghul** [comment]
> Any idea why is not ok 240 - fs_chown failing?

No idea, but doesn't seem related to your changes...

**Comment by saghul** [comment]
@refack Let's add the simple way to do it, and those who haven't updated to that version would need to call vcvarsall.bat themselves. @libuv/collaborators WDYT?

**Comment by refack** [comment]
I'll update today.

**Comment by zafields** [comment]
@refack Thanks for your work on this! We, the folks working on the [Azure IoT Gateway SDK](https://github.com/Azure/azure-iot-gateway-sdk/), are anxiously awaiting your fix. 👍

**Comment by zafields** [comment]
@saghul What is the status of this PR? It looks like @refack has made changes. I see you are still blocking, but I don't see any comments since the updates?

The [Azure IoT Edge SDK](https://github.com/azure/iot-edge) has a [bug](https://github.com/Azure/iot-edge/issues/218) (with work around) based on the LibUV dependency. It would be great to see this PR come in, we have a bunch of people looking forward to it.

Thanks,
Zak

**Comment by cjihrig** [comment]
Fresh CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/277/

@zafields have you tested the change in it's current state to verify that it fixes your problem? Unfortunately, the CI doesn't test VS2017.

**Comment by zafields** [comment]
@cjihrig To be honest, no I have not. We have a workaround in place and have moved forward.

The immediate problem we are experiencing, is that compiling with VS2017 isn't even a condition in the `if->else` checks, so it falls to the final else which attempts to use a very old version of the `vsbuild`. I am _assuming_ (I know, I'm embarrassed even writing the word) that once LibUV can be built with VS2017, then my problem will be solved. However, I am _certain_ that my problem will not be solved until LibUV is in fact buildable with VS2017.

Mostly, I just check in on this PR from time to time, because by it's name alone it looks to be the answer to my problem. To me, it looks like @refack has done the work, and the PR has just gone cold, so I wanted to ping the reviewers.

**Comment by refack** [comment]
> Mostly, I just check in on this PR from time to time, because by it's name alone it looks to be the answer to my problem. To me, it looks like @refack has done the work, and the PR has just gone cold, so I wanted to ping the reviewers.

Not exactly, current solution requires the dev to call `vcvars*.bat` manually. I promised to add a `vswhere` check before falling back to this. But then I got "sucked in" to the whirlwind that is being a `nodejs` Collaborator.

I'll push that now.

**Comment by refack** [comment]
@saghul Implemented `vswhere` detection with current behaviour as fall back.If `vswhere` is not found this is the output:
```console
'vswhere' is not recognized as an internal or external command,
operable program or batch file.
Warning: Visual Studio not found
['-Dtarget_arch=ia32', '-Duv_library=static_library', '.\\uv.gyp', '-I', '.\\common.gypi', '--depth=.', '-Dhost_arch=x64']
Project files generated.
Build skipped. To build, this file needs to run from VS cmd prompt.
```

**Comment by saghul** [comment]
CI (just in case) https://ci.nodejs.org/view/libuv/job/libuv-test-commit/280/

**Comment by schlamar** [comment]
Just tested it locally with VS 2017 Build Tools. Everything works as expected and tests pass 👍

**Comment by schlamar** [comment]
**Update** @saghul This will break pyuv builds! There is no way to skip building with VS 2017 if it is installed. So builds for Python < 3.5 won't work anymore as they need another VS compiler. 

Right now the correct VS compiler can be selected in pyuv by modifying the environment, the wrong `VSXXXCOMNTOOLS` variables are removed before calling vcbuild.bat ([ref](https://github.com/saghul/pyuv/blob/v1.x/setup_libuv.py#L65)).

So there should be a way to disable building with VS 2017 even if VS 2017 is installed.

**Comment by schlamar** [comment]
Something like

    if defined LIBUV_SKIP_VSWHERE_DETECTION goto :no-vs2017-env

**Comment by refack** [comment]
> Update @saghul This will break pyuv builds! There is no way to skip building with VS 2017 if it is installed. So builds for Python < 3.5 won't work anymore as they need another VS compiler.
> 
> Right now the correct VS compiler can be selected in pyuv by modifying the environment, the wrong VSXXXCOMNTOOLS variables are removed before calling vcbuild.bat (ref).
> 
> So there should be a way to disable building with VS 2017 even if VS 2017 is installed.

@schlamar 
I added a command line arg `skip2017`, does that help? You could also `set skip2017=1`, and it'll take.

**Comment by schlamar** [comment]
I would prefer a more specific environment variable, `skip2017` is **very** generally. 

As command line argument it'd OK. However, I'm not sure if that's useful as you can't skip other VS versions via command line arguments.

But maybe we should change the logic and instead of skipping a user can set a preferred VS version. We could use `GYP_MSVS_VERSION` for that. If that variable is defined we check only for the selected VS version. @saghul What do you think?

**Comment by schlamar** [comment]
Oh, my proposal with `GYP_MSVS_VERSION` won't work for pyuv as you can build Python 3.5+ extensions with VS 2015 **and** VS 2017. So you cannot select a preferred version. In this case skipping is the way to go. 

I would propose `LIBUV_SKIP_VS2017` as env variable. But maybe we should then support `LIBUV_SKIP_VSXXX` as well?

**Comment by refack** [comment]
> I would propose LIBUV_SKIP_VS2017 as env variable.

Ack.


> But maybe we should then support LIBUV_SKIP_VSXXX as well?

That I'm leaving for a different PR, since you need to decide on a fall through policy (yes/no? maybe the var should be a positive selection like `GYP_MSVS_VERSION` not `SKIP` logic?)

**Comment by refack** [comment]
Rebased

**Comment by saghul** [comment]
@schlamar Thanks for catching that! As you know we remove things from the environment when building on Windows in order to pick the right compiler. Won't this work if we skip VS 2017?

**Comment by saghul** [comment]
@refack @schlamar Let's go back a bit and re-think this. Adding a way to skip 2017 is just a hack IMHO. We should maybe read the GYP_MSVS_VERSION env variable and just try to use the requested one instead. In pyuv we'd set that.

Another approach I've been thinking about (for pyuv) is to ditch the libuv build system completely and add the libuv sources, similar to how I build pycares: https://github.com/saghul/pycares/blob/master/setup_cares.py that would simplify the build a lot, and remove Python 2 as a dependency when building on Python 3, among other things.

So, I suggest we go back to the previous scenario, fixup pyuv, and come back with a follow-up PR if we need it after the build refactor. This might be the excuse I was looking for in order to finally do it :-)

**Comment by schlamar** [comment]
@saghul Python 3.5+ extensions can be build with VS 2017 so we should skip VS 2017 only on Python < 3.5. For Python 3.5+ VS 2017 **or** 2015 should be allowed.

The current state of the PR should work for this use case. 

I wouldn't introduce a command line arg, though :)

**Comment by schlamar** [comment]
@saghul We cannot select a specific VS version, see my previous comment.

**Comment by saghul** [comment]
@schlamar Wouldn't https://github.com/libuv/libuv/pull/1284#issuecomment-305099237 solve it? It would be like building any other extension, since we wouldn't be using the libuv build system.

**Comment by schlamar** [comment]
Yes, I guess this would work and I really like your idea. 

A one liner `if defined LIBUV_SKIP_VS2017 goto ...` shouldn't hurt, though :)

**Comment by refack** [comment]
Removed "skip" logic.
I can help with a new PR implementing "positive selection" logic

**Comment by refack** [comment]
Removed "skip" logic.
I can help with a new PR implementing "positive selection" logic.

**Comment by ganeshmahajan1985** [comment]
Hi, I am trying to build but i am getting error. kindly check below details
My machine conf. win10 x64
VS 2017 enterprise
git version 2.13.0.windows.1
python 2.7.13

Please help to resolve this.

image



capture

**Comment by refack** [comment]
Try to run the build from a "Developer Command Prompt for VS 2017"
image

**Comment by schlamar** [comment]
@ganeshmahajan1985 It looks like you want to build v1.10.1 with VS 2017 which is not supported. You need to fetch from this PR (see https://help.github.com/articles/checking-out-pull-requests-locally/) if you want to build libuv with VS 2017.

**Comment by ganeshmahajan1985** [comment]
@schlamar  Interesting. 
Let me try this.
Thanks.
@refack Yes I am using developer console only. :)

**Comment by zafields** [comment]
Hi guys,

I am just following up. It looks like all green lights and no merge. We are getting about 3 issues a week based on lack of VS2017 support with libUV.

Do you have a timeline, any status or a short description of the hang-up I can pass along to our users?

Thanks,
Zak

**Comment by zafields** [comment]
@refack @schlamar @saghul Can you provide any insight on the status of this PR?

**Comment by bzoz** [comment]
@schlamar @refack vs2017 support would be nice, can we make it opt-in for the time being, before pyuv gets a fix? Maybe an command line arg, like `vcbuild vc2017` ?

**Comment by refack** [comment]
Done, PTAL.
I used as much code from `nodejs` as was reasonable so it's a tad bit noisy, but IMHO it's best
I also updated the helper

File: tools/vswhere_usability_wrapper.cmd [hunk 18]
```diff
@@ -0,0 +1,33 @@
+:: Copyright 2017 - Refael Ackermann
+:: Distributed under MIT style license or the libuv license
+:: See accompanying file LICENSE at https://github.com/node4good/windows-autoconf
+:: or libuv LICENSE file at https://github.com/libuv/libuv
+:: version: 1.15.0
+
+@if not defined DEBUG_HELPER @ECHO OFF
+setlocal
+set "InstallerPath=%ProgramFiles(x86)%\Microsoft Visual Studio\Installer"
+if not exist "%InstallerPath%" set "VSWHERE=%ProgramFiles%\Microsoft Visual Studio\Installer"
```
Comment (bzoz) [inline tools/vswhere_usability_wrapper.cmd]: Shouldn't this be set `"InstallerPath=..."`?

File: vcbuild.bat [hunk 19]
```diff
@@ -41,10 +43,29 @@ shift
 goto next-arg
 :args-done
 
-if defined WindowsSDKDir goto select-target
-if defined VCINSTALLDIR goto select-target
+@rem Look for Visual Studio 2017
+:vs-set-2017
```
Comment (bzoz) [inline vcbuild.bat]: this label is unused

**Comment by bzoz** [comment]
Also, since things changed quite some: @cjihrig @bnoordhuis @saghul can you reevaluate your reviews?

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/307/

File: tools/vswhere_usability_wrapper.cmd (hunk 18 above)
Comment (refack) [inline tools/vswhere_usability_wrapper.cmd]: Thanks! I've fixed it upstream as well.

File: vcbuild.bat (hunk 19 above)
Comment (refack) [inline vcbuild.bat]: It was there just for symmetry 🤷‍♂️ . Removed, we'll bring it back for VS2019 😉

**Comment by refack** [comment]
Well It's building, so I assume CI fail on `win2008r2-vs2013` is unrelated?
```
not ok 282 - threadpool_cancel_getnameinfo
# exit code 3
# Output from process `threadpool_cancel_getnameinfo`:
# A
```
https://ci.nodejs.org/job/libuv-test-commit-windows/371/nodes=win2008r2-vs2013/

**Comment by bzoz** [comment]
I guess so.

File: vcbuild.bat [hunk 20]
```diff
@@ -29,6 +30,7 @@ if /i "%1"=="release"      set config=Release&goto arg-ok
 if /i "%1"=="test"         set run=run-tests.exe&goto arg-ok
 if /i "%1"=="bench"        set run=run-benchmarks.exe&goto arg-ok
 if /i "%1"=="clean"        set target=Clean&goto arg-ok
+if /i "%1"=="vs2017"        set target_env=vs2017&goto arg-ok
```
Comment (bnoordhuis) [inline vcbuild.bat]: Minor indent error.

Comment (refack) [inline vcbuild.bat]: Ack

File: vcbuild.bat [hunk 21]
```diff
@@ -14,10 +14,11 @@ if /i "%1"=="/?" goto help
 @rem Process arguments.
 set config=
 set target=Build
+set target_arch=ia32
+set target_env=vs2015
```
Comment (saghul) [inline vcbuild.bat]: can we call this vslegacy or something like that instead? at a glace it looks like I could set it to vs2013 if I wanted that, but that's not the case.

Comment (refack) [inline vcbuild.bat]: I simply removed the default value (the downstream code test works as expected with an empty or undefined value anyway)

Comment (saghul) [inline vcbuild.bat]: Looks good! Some docs and we're good to go here.

**Comment by refack** [comment]
@saghul I could find any reference to VS building except for https://github.com/libuv/libuv#windows and it still holds...

**Comment by saghul** [comment]
@refack Time to extend it then ;-)

**Comment by refack** [comment]
> @refack Time to extend it then ;-)

How much? do you want whole blurb from `node` https://github.com/nodejs/node/blob/master/BUILDING.md#windows-1?

**Comment by saghul** [comment]
> How much? do you want whole blurb from node https://github.com/nodejs/node/blob/master/BUILDING.md#windows-1?

Yeah, that would be great! Thanks! 🙏

**Comment by refack** [comment]
Added, adjusted, and de-linted (based on https://github.com/nodejs/node/pull/12756)

**Comment by saghul** [comment]
Thanks for your perseverance and patience! Landed in 19c896dd086abb351361eab54b7e88ed5ce8af08, 0953bf0dad50d0dc77fec2deefe1c5b6c5c3f020 and 80d3b95f57e4509d93b3470e60d090b0cddb216a. 🎉

**Comment by cjihrig** [comment]
@refack is this still building successfully for you on AppVeyor? It doesn't seem like it based on https://ci.appveyor.com/project/refack/libuv/build/v1.12.0.build41. That looks like the same error I'm currently seeing while trying to cut the 1.13.0 release.

**Comment by refack** [comment]
> @refack is this still building successfully for you on AppVeyor? It doesn't seem like it based on https://ci.appveyor.com/project/refack/libuv/build/v1.12.0.build41. That looks like the same error I'm currently seeing while trying to cut the 1.13.0 release.

Havn't looked at it in a while, looking now.
